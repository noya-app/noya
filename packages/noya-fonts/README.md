# Noya Fonts

A utility for managing fonts
