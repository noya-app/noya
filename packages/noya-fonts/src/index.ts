import { Brand } from 'noya-utils';

export * from './fontTraits';
export * from './fontDescriptor';
export * from './fontManager';
export * from './fontWeight';

export function formatFontFamilyId(fontFamily: string) {
  return fontFamily.toLowerCase().replace(/[ _-]/g, '');
}

export type FontFamilyId = Brand<string, 'fontFamily'>;
