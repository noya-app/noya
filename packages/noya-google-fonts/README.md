# Noya Google Fonts

Utilities for working with Google Fonts. We include a pre-downloaded list of
fonts so that they can be ready as soon as possible.
