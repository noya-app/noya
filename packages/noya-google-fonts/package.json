{
  "name": "noya-google-fonts",
  "version": "1.0.0",
  "main": "src/index.ts",
  "license": "UNLICENSED"
}
