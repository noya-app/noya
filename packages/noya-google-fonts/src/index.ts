const webfontList: WebfontList = require('./fonts.json');
const fonts = webfontList.items;
const fontMap = fonts.reduce((result: Record<string, WebfontFamily>, item) => {
  result[item.family.toLowerCase()] = item;
  return result;
}, {});

type FontCategory =
  | 'sans-serif'
  | 'serif'
  | 'display'
  | 'handwriting'
  | 'monospace';

type FontSubset =
  | 'latin'
  | 'latin-ext'
  | 'sinhala'
  | 'greek'
  | 'kannada'
  | 'telugu'
  | 'vietnamese'
  | 'hebrew'
  | 'cyrillic'
  | 'cyrillic-ext'
  | 'greek-ext'
  | 'arabic'
  | 'devanagari'
  | 'khmer'
  | 'tamil'
  | 'thai'
  | 'bengali'
  | 'gujarati'
  | 'oriya'
  | 'malayalam'
  | 'gurmukhi'
  | 'korean'
  | 'japanese'
  | 'tibetan'
  | 'chinese-simplified'
  | 'chinese-hongkong'
  | 'chinese-traditional'
  | 'myanmar';

export type RegularFontVariant =
  | '100'
  | '200'
  | '300'
  | 'regular'
  | '500'
  | '600'
  | '700'
  | '800'
  | '900';

export type ItalicFontVariant =
  | '100italic'
  | '200italic'
  | '300italic'
  | 'italic'
  | '500italic'
  | '600italic'
  | '700italic'
  | '800italic'
  | '900italic';

export type FontVariant = RegularFontVariant | ItalicFontVariant;

interface WebfontFamily {
  family: string;
  variants: FontVariant[];
  subsets: FontSubset[];
  version: string;
  lastModified: string;
  files: Record<FontVariant, string>;
  category: FontCategory;
  kind: 'webfonts#webfont';
}

interface WebfontList {
  items: WebfontFamily[];
  kind: 'webfonts#webfontList';
}

export function getFontFamilyList() {
  return fonts.map((font) => font.family);
}

export function hasFontFamily(fontFamily: string) {
  return getKey(fontFamily) in fontMap;
}

export type FontVariantCollection = {
  regular: RegularFontVariant[];
  italic: ItalicFontVariant[];
};

function isItalicVariant(variant: FontVariant): variant is ItalicFontVariant {
  return variant.includes('italic');
}

function isRegularVariant(variant: FontVariant): variant is RegularFontVariant {
  return !variant.includes('italic');
}

export function getFontVariants(fontFamily: string) {
  return fontMap[getKey(fontFamily)].variants;
}

export function getFontVariantCollection(
  fontFamily: string,
): FontVariantCollection {
  const variants = fontMap[getKey(fontFamily)].variants;
  const regular = variants.filter(isRegularVariant);
  const italic = variants.filter(isItalicVariant);

  return { regular, italic };
}

export function getFontFile(fontFamily: string, variant: FontVariant) {
  return fontMap[getKey(fontFamily)].files[variant];
}

function getKey(fontFamily: string) {
  return fontFamily.toLowerCase();
}

type FontWeight =
  | 'ultralight'
  | 'thin'
  | 'light'
  | 'regular'
  | 'medium'
  | 'semibold'
  | 'bold'
  | 'heavy'
  | 'black';

export function getFontVariantWeight(variant: FontVariant) {
  switch (variant) {
    case '100':
    case '100italic':
      return 'ultralight';
    case '200':
    case '200italic':
      return 'thin';
    case '300':
    case '300italic':
      return 'light';
    case 'regular':
    case 'italic':
      return 'regular';
    case '500':
    case '500italic':
      return 'medium';
    case '600':
    case '600italic':
      return 'semibold';
    case '700':
    case '700italic':
      return 'bold';
    case '800':
    case '800italic':
      return 'heavy';
    case '900':
    case '900italic':
      return 'black';
  }
}

export function isValidFontVariant(string: string): string is FontVariant {
  switch (string) {
    case '100':
    case '100italic':
    case '200':
    case '200italic':
    case '300':
    case '300italic':
    case 'regular':
    case 'italic':
    case '500':
    case '500italic':
    case '600':
    case '600italic':
    case '700':
    case '700italic':
    case '800':
    case '800italic':
    case '900':
    case '900italic':
      return true;
    default:
      return false;
  }
}

export function decodeFontVariant(
  variant: FontVariant,
): {
  weight: FontWeight;
  variantName: 'regular' | 'italic';
} {
  return {
    weight: getFontVariantWeight(variant),
    variantName: isItalicVariant(variant) ? 'italic' : 'regular',
  };
}
