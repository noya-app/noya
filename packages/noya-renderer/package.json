{
  "name": "noya-renderer",
  "version": "1.0.0",
  "main": "src/index.ts",
  "license": "UNLICENSED",
  "dependencies": {
    "@sketch-hq/sketch-file-format-ts": "^6.0.3",
    "canvaskit-wasm": "0.25.0",
    "uuid": "^8.3.2"
  },
  "devDependencies": {
    "@types/uuid": "^8.3.0",
    "canvas": "^2.6.1"
  }
}
