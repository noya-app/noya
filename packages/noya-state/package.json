{
  "name": "noya-state",
  "version": "1.0.0",
  "main": "src/index.ts",
  "license": "UNLICENSED",
  "dependencies": {
    "immer": "^9.0.5",
    "tree-visit": "^0.4.0"
  }
}
