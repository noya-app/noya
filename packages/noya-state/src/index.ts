import { setAutoFreeze } from 'immer';

setAutoFreeze(false);

export * as Overrides from './overrides';
export * as Layers from './layers';
export * as GroupLayouts from './groupLayouts';
export * as ExportOptions from './exportOptions';
export * as Selectors from './selectors/selectors';
export * as Models from './models';
export * as TextStyleSelectors from './selectors/textStyleSelectors';

export type { SelectedPoint } from './reducers/pointReducer';
export * from './reducers/workspaceReducer';
export * from './reducers/applicationReducer';
export * from './reducers/historyReducer';
export * from './reducers/interactionReducer';
export * from './types';

export type { SelectionType } from './utils/selection';

export * as Primitives from './primitives';
export * from './primitives';
export * from './selection';
export * from './checkeredBackground';
export * from './snapping';
