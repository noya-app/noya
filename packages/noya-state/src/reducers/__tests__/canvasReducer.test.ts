import type { CanvasKit as CanvasKitType } from 'canvaskit';
import { FontManager } from 'noya-fonts';
import { GoogleFontProvider } from 'noya-google-fonts';
import { loadCanvasKit } from 'noya-renderer';
import { debugDescription, SketchModel } from 'noya-sketch-model';
import { createInitialState, createSketchFile, Selectors } from 'noya-state';
import {
  encodePageMetadata,
  getCurrentPageMetadata,
} from '../../selectors/pageSelectors';
import {
  Action,
  applicationReducer,
  ApplicationReducerContext,
  ApplicationState,
} from '../applicationReducer';

let CanvasKit: CanvasKitType;
let context: ApplicationReducerContext;

beforeAll(async () => {
  CanvasKit = await loadCanvasKit();
  const typefaceFontProvider = CanvasKit.TypefaceFontProvider.Make();
  context = {
    canvasInsets: { top: 0, right: 0, bottom: 0, left: 0 },
    canvasSize: { width: 1000, height: 1000 },
    fontManager: {
      ...new FontManager(GoogleFontProvider),
      getTypefaceFontProvider: () => typefaceFontProvider,
    },
  };
});

function run(state: ApplicationState, actions: Action[]) {
  return actions.reduce(
    (result, action) => applicationReducer(result, action, CanvasKit, context),
    state,
  );
}

const rectangle = SketchModel.rectangle({
  frame: SketchModel.rect({
    x: 0,
    y: 0,
    width: 100,
    height: 100,
  }),
});

const oval = SketchModel.oval({
  frame: SketchModel.rect({
    x: 200,
    y: 200,
    width: 100,
    height: 100,
  }),
});

describe('move', () => {
  test('move one layer', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeMove', { x: 10, y: 10 }]],
      ['interaction', ['updateMoving', { x: 25, y: 25 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('move multiple layers', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, oval] })),
    );
    state.selectedObjects = [rectangle.do_objectID, oval.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeMove', { x: 10, y: 10 }]],
      ['interaction', ['updateMoving', { x: 25, y: 25 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('move with snap', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, oval] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const mouseOriginX = 10;

    const updated = run(state, [
      ['interaction', ['maybeMove', { x: mouseOriginX, y: 10 }]],
      [
        'interaction',
        ['updateMoving', { x: oval.frame.x + mouseOriginX - 3, y: 25 }],
      ],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('move layer in artboard', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 450,
        y: 450,
        width: 300,
        height: 300,
      }),
      layers: [rectangle],
    });

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeMove', { x: 460, y: 460 }]],
      ['interaction', ['updateMoving', { x: 475, y: 475 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });
});

describe('movingPoint', () => {
  test('move one point', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['editPath']],
      ['selectPoint', [rectangle.do_objectID, 0]],
      ['interaction', ['maybeMovePoint', { x: 0, y: 0 }]],
      ['interaction', ['movingPoint', { x: 0, y: 0 }, { x: 25, y: 25 }]],
    ]);

    expect(
      debugDescription(
        [Selectors.getCurrentPage(state), Selectors.getCurrentPage(updated)],
        { points: true },
      ),
    ).toMatchSnapshot();
  });

  test('move point in artboard', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 100,
        y: 100,
        width: 300,
        height: 300,
      }),
      layers: [rectangle],
    });

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard] })),
    );

    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['editPath']],
      ['selectPoint', [rectangle.do_objectID, 0]],
      ['interaction', ['maybeMovePoint', { x: 0, y: 0 }]],
      ['interaction', ['movingPoint', { x: 0, y: 0 }, { x: 25, y: 25 }]],
    ]);

    expect(
      debugDescription(
        [Selectors.getCurrentPage(state), Selectors.getCurrentPage(updated)],
        { points: true },
      ),
    ).toMatchSnapshot();
  });
});

describe('movingControlPoint', () => {
  test('move one control point', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [oval] })),
    );
    state.selectedObjects = [oval.do_objectID];

    const updated = run(state, [
      ['interaction', ['editPath']],
      ['selectControlPoint', oval.do_objectID, 0, 'curveFrom'],
      ['interaction', ['maybeMoveControlPoint', { x: 0, y: 0 }]],
      ['interaction', ['movingControlPoint', { x: 0, y: 0 }, { x: 25, y: 25 }]],
    ]);

    expect(
      debugDescription(
        [Selectors.getCurrentPage(state), Selectors.getCurrentPage(updated)],
        { points: true },
      ),
    ).toMatchSnapshot();
  });

  test('move control point in artboard', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 100,
        y: 100,
        width: 300,
        height: 300,
      }),
      layers: [oval],
    });

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard] })),
    );

    state.selectedObjects = [oval.do_objectID];

    const updated = run(state, [
      ['interaction', ['editPath']],
      ['selectControlPoint', oval.do_objectID, 0, 'curveFrom'],
      ['interaction', ['maybeMoveControlPoint', { x: 0, y: 0 }]],
      ['interaction', ['movingControlPoint', { x: 0, y: 0 }, { x: 25, y: 25 }]],
    ]);

    expect(
      debugDescription(
        [Selectors.getCurrentPage(state), Selectors.getCurrentPage(updated)],
        { points: true },
      ),
    ).toMatchSnapshot();
  });
});

describe('scale', () => {
  test('scale one layer se', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeScale', { x: 100, y: 100 }, 'se']],
      ['interaction', ['updateScaling', { x: 125, y: 125 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('scale one layer ne', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeScale', { x: 100, y: 100 }, 'ne']],
      ['interaction', ['updateScaling', { x: 125, y: 125 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('scale one layer nw', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeScale', { x: 100, y: 100 }, 'nw']],
      ['interaction', ['updateScaling', { x: 125, y: 125 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('scale one layer sw', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeScale', { x: 100, y: 100 }, 'sw']],
      ['interaction', ['updateScaling', { x: 125, y: 125 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('scale multiple layer', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, oval] })),
    );
    state.selectedObjects = [rectangle.do_objectID, oval.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeScale', { x: 100, y: 100 }, 'se']],
      ['interaction', ['updateScaling', { x: 125, y: 125 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('scale one layer in artboard', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 450,
        y: 450,
        width: 300,
        height: 300,
      }),
      layers: [rectangle],
    });
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard] })),
    );
    state.selectedObjects = [rectangle.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeScale', { x: 100, y: 100 }, 'se']],
      ['interaction', ['updateScaling', { x: 125, y: 125 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('scale layers in different parents', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 450,
        y: 450,
        width: 300,
        height: 300,
      }),
      layers: [rectangle],
    });
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard, oval] })),
    );
    state.selectedObjects = [rectangle.do_objectID, oval.do_objectID];

    const updated = run(state, [
      ['interaction', ['maybeScale', { x: 100, y: 100 }, 'se']],
      ['interaction', ['updateScaling', { x: 125, y: 125 }]],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });
});

describe('drawing', () => {
  test('draw rectangle layer', () => {
    const state = createInitialState(createSketchFile(SketchModel.page()));

    const updated = run(state, [
      ['interaction', ['startDrawing', 'rectangle', { x: 0, y: 0 }]],
      ['interaction', ['updateDrawing', { x: 100, y: 100 }]],
      ['addDrawnLayer'],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('draw oval layer', () => {
    const state = createInitialState(createSketchFile(SketchModel.page()));

    const updated = run(state, [
      ['interaction', ['startDrawing', 'oval', { x: 0, y: 0 }]],
      ['interaction', ['updateDrawing', { x: 100, y: 100 }]],
      ['addDrawnLayer'],
    ]);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('draw line layer', () => {
    const state = createInitialState(createSketchFile(SketchModel.page()));

    const updated = run(state, [
      ['interaction', ['startDrawing', 'line', { x: 0, y: 0 }]],
      ['interaction', ['updateDrawing', { x: 100, y: 100 }]],
      ['addDrawnLayer'],
    ]);

    expect(
      debugDescription(
        [Selectors.getCurrentPage(state), Selectors.getCurrentPage(updated)],
        { points: true },
      ),
    ).toMatchSnapshot();
  });
});

describe('setZoom', () => {
  test('zoom 2x', () => {
    const state = createInitialState(createSketchFile(SketchModel.page()));

    expect(getCurrentPageMetadata(state)).toEqual({
      zoomValue: 1,
      scrollOrigin: { x: 0, y: 0 },
    });

    const updated = run(state, [['setZoom', 2, 'multiply']]);

    expect(getCurrentPageMetadata(updated)).toEqual({
      zoomValue: 2,
      scrollOrigin: { x: -500, y: -500 },
    });
  });

  test('zoom 4x', () => {
    const page = SketchModel.page();
    const state = createInitialState(createSketchFile(page));

    state.sketch.user[page.do_objectID] = encodePageMetadata({
      zoomValue: 2,
      scrollOrigin: { x: -500, y: -500 },
    });

    const updated = run(state, [['setZoom', 2, 'multiply']]);

    expect(getCurrentPageMetadata(updated)).toEqual({
      zoomValue: 4,
      scrollOrigin: { x: -1500, y: -1500 },
    });
  });

  test('zoom 0.5x', () => {
    const page = SketchModel.page();
    const state = createInitialState(createSketchFile(page));

    state.sketch.user[page.do_objectID] = encodePageMetadata({
      zoomValue: 2,
      scrollOrigin: { x: -500, y: -500 },
    });

    const updated = run(state, [['setZoom', 1 / 2, 'multiply']]);

    expect(getCurrentPageMetadata(updated)).toEqual({
      zoomValue: 1,
      scrollOrigin: { x: 0, y: 0 },
    });
  });

  test('zoom when no user metadata exists for a page', () => {
    const state = createInitialState(createSketchFile());
    const page = SketchModel.page();
    state.sketch.pages = [page];
    state.selectedPage = page.do_objectID;

    expect(state.sketch.user[page.do_objectID]).toEqual(undefined);

    const updated = run(state, [['setZoom', 2, 'multiply']]);

    expect(getCurrentPageMetadata(updated)).toEqual({
      zoomValue: 2,
      scrollOrigin: { x: -500, y: -500 },
    });
  });
});
