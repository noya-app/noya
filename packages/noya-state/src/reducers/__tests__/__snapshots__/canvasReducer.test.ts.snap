// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawing draw line layer 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <ShapePath> Path { x: 0, y: 0, w: 100, h: 100 }
    ├── <CurvePoint> { x: 0, y: 0 } (Straight)
    └── <CurvePoint> { x: 1, y: 1 } (Straight)"
`;

exports[`drawing draw oval layer 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Oval> Oval { x: 0, y: 0, w: 100, h: 100 }"
`;

exports[`drawing draw rectangle layer 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }"
`;

exports[`move move layer in artboard 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 450, y: 450, w: 300, h: 300 }
    └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 450, y: 450, w: 300, h: 300 }
    └── <Rectangle> Rectangle { x: 15, y: 15, w: 100, h: 100 }"
`;

exports[`move move multiple layers 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval { x: 200, y: 200, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 15, y: 15, w: 100, h: 100 }
└── <Oval> Oval { x: 215, y: 215, w: 100, h: 100 }"
`;

exports[`move move one layer 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 15, y: 15, w: 100, h: 100 }"
`;

exports[`move move with snap 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval { x: 200, y: 200, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 200, y: 15, w: 100, h: 100 }
└── <Oval> Oval { x: 200, y: 200, w: 100, h: 100 }"
`;

exports[`movingControlPoint move control point in artboard 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 100, y: 100, w: 300, h: 300 }
    └── <Oval> Oval { x: 200, y: 200, w: 100, h: 100 }
        ├── <CurvePoint> { x: 0.5, y: 1 } (Mirrored)
        ├── <CurvePoint> { x: 1, y: 0.5 } (Mirrored)
        ├── <CurvePoint> { x: 0.5, y: 0 } (Mirrored)
        └── <CurvePoint> { x: 0, y: 0.5 } (Mirrored)

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 100, y: 100, w: 300, h: 300 }
    └── <Oval> Oval { x: 200, y: 200, w: 100, h: 107.05 }
        ├── <CurvePoint> { x: 0.5, y: 0.93 } (Mirrored)
        ├── <CurvePoint> { x: 1, y: 0.47 } (Mirrored)
        ├── <CurvePoint> { x: 0.5, y: 0 } (Mirrored)
        └── <CurvePoint> { x: 0, y: 0.47 } (Mirrored)"
`;

exports[`movingControlPoint move one control point 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Oval> Oval { x: 200, y: 200, w: 100, h: 100 }
    ├── <CurvePoint> { x: 0.5, y: 1 } (Mirrored)
    ├── <CurvePoint> { x: 1, y: 0.5 } (Mirrored)
    ├── <CurvePoint> { x: 0.5, y: 0 } (Mirrored)
    └── <CurvePoint> { x: 0, y: 0.5 } (Mirrored)

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Oval> Oval { x: 200, y: 200, w: 100, h: 107.05 }
    ├── <CurvePoint> { x: 0.5, y: 0.93 } (Mirrored)
    ├── <CurvePoint> { x: 1, y: 0.47 } (Mirrored)
    ├── <CurvePoint> { x: 0.5, y: 0 } (Mirrored)
    └── <CurvePoint> { x: 0, y: 0.47 } (Mirrored)"
`;

exports[`movingPoint move one point 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
    ├── <CurvePoint> { x: 0, y: 0 } (Straight)
    ├── <CurvePoint> { x: 1, y: 0 } (Straight)
    ├── <CurvePoint> { x: 1, y: 1 } (Straight)
    └── <CurvePoint> { x: 0, y: 1 } (Straight)

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
    ├── <CurvePoint> { x: 0.25, y: 0.25 } (Straight)
    ├── <CurvePoint> { x: 1, y: 0 } (Straight)
    ├── <CurvePoint> { x: 1, y: 1 } (Straight)
    └── <CurvePoint> { x: 0, y: 1 } (Straight)"
`;

exports[`movingPoint move point in artboard 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 100, y: 100, w: 300, h: 300 }
    └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
        ├── <CurvePoint> { x: 0, y: 0 } (Straight)
        ├── <CurvePoint> { x: 1, y: 0 } (Straight)
        ├── <CurvePoint> { x: 1, y: 1 } (Straight)
        └── <CurvePoint> { x: 0, y: 1 } (Straight)

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 100, y: 100, w: 300, h: 300 }
    └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
        ├── <CurvePoint> { x: 0.25, y: 0.25 } (Straight)
        ├── <CurvePoint> { x: 1, y: 0 } (Straight)
        ├── <CurvePoint> { x: 1, y: 1 } (Straight)
        └── <CurvePoint> { x: 0, y: 1 } (Straight)"
`;

exports[`scale scale layers in different parents 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Artboard> Artboard { x: 450, y: 450, w: 300, h: 300 }
│   └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval { x: 200, y: 200, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Artboard> Artboard { x: 450, y: 450, w: 300, h: 300 }
│   └── <Rectangle> Rectangle { x: 18, y: 18, w: 107, h: 107 }
└── <Oval> Oval { x: 200, y: 200, w: 107, h: 107 }"
`;

exports[`scale scale multiple layer 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval { x: 200, y: 200, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 108, h: 108 }
└── <Oval> Oval { x: 217, y: 217, w: 108, h: 108 }"
`;

exports[`scale scale one layer in artboard 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 450, y: 450, w: 300, h: 300 }
    └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 450, y: 450, w: 300, h: 300 }
    └── <Rectangle> Rectangle { x: 0, y: 0, w: 125, h: 125 }"
`;

exports[`scale scale one layer ne 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 25, w: 125, h: 75 }"
`;

exports[`scale scale one layer nw 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 25, y: 25, w: 75, h: 75 }"
`;

exports[`scale scale one layer se 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 125, h: 125 }"
`;

exports[`scale scale one layer sw 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 25, y: 0, w: 75, h: 125 }"
`;
